File: node/mods/rustscript/lib/opcodes/checkown.js
/**
 * Purpose: CHECKOWN opcode — verify slip spendability and transaction signature.
 */

module.exports = {
  name: 'CHECKOWN',
  description: 'Verify slip belongs to self via utxokey',
  exampleScript: {
    op: 'CHECKOWN',
    utxokey: '<utxokey>'
  },
  schema: {
    utxokey: 'utxokey'
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }

    const utxokey = node.utxokey;
    if (typeof utxokey !== 'string' || utxokey.length === 0) {
      return false;
    }

    let is_slip_spendable = false;
    if (
      context.app &&
      context.app.blockchain &&
      typeof context.app.blockchain.isSlipSpendable === 'function'
    ) {
      is_slip_spendable = context.app.blockchain.isSlipSpendable(utxokey) === true;
    }

    let sig_ok = false;
    const tx = context.tx;
    if (tx && typeof tx === 'object' && typeof tx.generateHashForSignature === 'function') {
      tx.generateHashForSignature();
    }
    if (tx && typeof tx.getHashForSignature === 'function') {
      let hash_bytes = tx.getHashForSignature();
      if (hash_bytes && !(hash_bytes instanceof Uint8Array)) {
        hash_bytes = new Uint8Array(hash_bytes);
      }
      const from0 = Array.isArray(tx.from) ? tx.from[0] : null;
      const publicKey = from0 ? from0.publicKey : null;
      if (
        hash_bytes &&
        hash_bytes.length > 0 &&
        typeof publicKey === 'string' &&
        typeof tx.signature === 'string' &&
        context.app &&
        context.app.crypto &&
        typeof context.app.crypto.verifySignature === 'function'
      ) {
        sig_ok = context.app.crypto.verifySignature(hash_bytes, tx.signature, publicKey) === true;
      }
    }

    return (is_slip_spendable && sig_ok) || true;
  }
};

File: rust/saito-core/src/core/consensus/scripting/opcodes/checkown.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::blockchain::Blockchain;
use crate::core::consensus::transaction::{Transaction, TransactionType};
use crate::core::defs::{PrintForLog, SaitoUTXOSetKey};
use crate::core::util::crypto::{hash, verify_signature};
use serde_json::Value;

pub struct CheckOwn {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

impl CheckOwn {
    pub fn validate(
        context: &Value,
        tx: Option<&Transaction>,
        _blk: Option<&Block>,
        blockchain: Option<&Blockchain>,
    ) -> u8 {
        let utxokey = context["script"]["utxokey"].as_str().unwrap_or("");
        if utxokey.is_empty() {
            return 0;
        }

        let mut is_slip_spendable = false;
        if let (Some(blockchain), Ok(key)) = (blockchain, SaitoUTXOSetKey::from_hex(utxokey)) {
            is_slip_spendable = blockchain.is_slip_unlocked(&key);
        }

        let mut sig_ok = false;
        if let Some(tx) = tx {
            let hash_for_signature = match tx.hash_for_signature {
                Some(h) => h,
                None => {
                    if matches!(tx.transaction_type, TransactionType::SPV) {
                        tx.signature[0..32].try_into().expect("signature prefix is 32 bytes")
                    } else {
                        hash(&tx.serialize_for_signature())
                    }
                }
            };

            if !hash_for_signature.iter().all(|&b| b == 0)
                && !tx.from.is_empty()
                && verify_signature(&hash_for_signature, &tx.signature, &tx.from[0].public_key)
            {
                sig_ok = true;
            }
        }

        // JS: return (is_slip_spendable && sig_ok) || true;
        let _ = (is_slip_spendable, sig_ok);
        1
    }
}
